.github/scripts/compute_and_upload_ci_wait_time_metric.py

```diff
@@ -0,0 +1,389 @@
+        ),
+    ]
+
+    data = []
+    page_num = 1
+    start_time = time.time()
+    for page in rockset.QueryPaginator(
+        rs,
+        rs.QueryLambdas.execute_query_lambda(
+            query_lambda=queryLambdaName,
+            version=version,
+            workspace=collection,
+            parameters=params,
+            paginate=True,
+            initial_paginate_response_doc_count=10000,
+        )
+    ):
+        print(f"Got page {page_num}")
+        page_num += 1
+        data += page
+
+    print(f"Total rows received: {len(data)}")
+    end_time = time.time()
+    print(f"Total query time: {end_time - start_time} seconds ")
+
+    results_df = pd.json_normalize(data)
+
+    return normalize_workflow_runs(results_df)
+
+
+def remove_cancelled_jobs(df):
+    # Cancelled jobs can fall into one of three categories:
+    # 1. The job timed out
+    # 2. The run was cancelled by a user's push before any job failed
+    # 3. The run was cancelled by a user's push after a job failed
+
+    # Identify the cancellations that were timeouts. It's a timeout if the job was cancelled after 5hrs,
+    df['was_timeout'] = df.apply(
+        lambda row: row['was_cancelled'] and row['duration_mins'] >= 300, axis=1)
+
+    # Consider timeouts to be a type of failures.
+    df['conclusion'] = df.apply(
+        lambda row: 'failure' if row['was_timeout'] else row['conclusion'], axis=1)
+
+    # For non-timeout cancellations, if the workflows were cancelled _after_ a job already failed, then
+    # we say the failure signal gave the "end time" since it provided an actionable signal.
+    # This means we can ignore the 'cancelled' jobs for such workflow runs.
+    # Thus, if the job failed, and later it's workflow was cancelled then we ignore those cancelled jobs
+    df = df[~((df['conclusion'] == 'cancelled') & (
+        df['sha'].isin(df[df['conclusion'] == 'failure']['sha'])))]
+
+    # Now the remaining cancelled jobs are the ones cancelled due to additional commits being pushed
+    # _before_ a job failed. We can ignore these entire commits since they did't provide any actionable signal.
+    # Thus we drop all shas that still contain a cancelled job
+    df = df[~df['sha'].isin(df[df['conclusion'] == 'cancelled']['sha'])]
+
+    return df
+
+
+# For each workflow run, set the start_time for the group to the earliest start_time in that group
+# since the first job shows when the user started waiting on CI
+def normalize_start_times(df):
+    # We track each run attempt in a workflow run separately.
+    df_grouped = df.groupby(['sha', 'run_attempt', 'workflow_run_id'])
+
+    # Suppress warning "FutureWarning: Passing 'suffixes' which cause duplicate columns {'start_time_orig'} in the result is deprecated and will raise a MergeError in a future version."
+    with warnings.catch_warnings():
+        warnings.simplefilter(action='ignore', category=FutureWarning)
+        df = df.merge(df_grouped['start_time'].min(), on=[
```

Comment:
IIRC, you could use the workflow `start_time` here instead of job `start_time` from the query `min(j._event_time) AS start_time`.  That's how the TTS query is done at the moment:

* TTS = job.completed_at - workflow.event_time
* Duration = job.complete_at - job.event_time

This is under the assumption that the event time of the first job would roughly be the same as the workflow event time (unless something goes horribly wrong with infra)

Reply:
Nice. Changed the query to use `workflow_run.run_started_at`. Skimming through the results, it pulls the start time back about 10-20 seconds, which sounds like a reasonable diff from the original version.